File: easycontrol/adb/src/main/java/top/saymzx/adb/Adb.java
package top.saymzx.adb;

import java.io.IOException;
import java.io.InputStream;
import java.nio.ByteBuffer;
import java.util.concurrent.ConcurrentHashMap;

import top.saymzx.buffer.BufferNew;
import top.saymzx.buffer.Stream;

// 此部分代码摘抄借鉴了tananaev大佬的开源代码(https://github.com/tananaev/adblib)以及开源库dadb(https://github.com/mobile-dev-inc/dadb)
public class Adb {
  private final AdbChannel channel;
  private int localIdPool = 1;
  private final ConcurrentHashMap<Integer, Stream> connectionStreams = new ConcurrentHashMap<>();
  private final BufferNew sendBuffer = new BufferNew();

  private final Thread handleInThread = new Thread(this::handleIn);
  private final Thread handleOutThread = new Thread(this::handleOut);

  public Adb(String host, int port, AdbKeyPair keyPair) throws Exception {
    channel = new TcpChannel(host, port);
    connect(keyPair);
  }

  public Adb(Stream stream, AdbKeyPair keyPair) throws Exception {
    channel = new StreamChannel(stream);
    connect(keyPair);
  }

  private void connect(AdbKeyPair keyPair) throws Exception {
    // 授权超时
    Thread startListener = new Thread(() -> {
      try {
        Thread.sleep(10 * 1000);
        close();
      } catch (InterruptedException ignored) {
      }
    });
    startListener.start();
    // 连接ADB并认证
    channel.write(AdbProtocol.generateConnect().array());
    AdbProtocol.AdbMessage message = AdbProtocol.AdbMessage.parseAdbMessage(channel);
    if (message.command == AdbProtocol.CMD_AUTH) {
      channel.write(AdbProtocol.generateAuth(AdbProtocol.AUTH_TYPE_SIGNATURE, keyPair.signPayload(message)).array());
      message = AdbProtocol.AdbMessage.parseAdbMessage(channel);
      if (message.command == AdbProtocol.CMD_AUTH) {
        channel.write(AdbProtocol.generateAuth(AdbProtocol.AUTH_TYPE_RSA_PUBLIC, keyPair.publicKeyBytes).array());
        message = AdbProtocol.AdbMessage.parseAdbMessage(channel);
      }
    }
    if (message.command != AdbProtocol.CMD_CNXN) throw new Exception("ADB连接失败: " + message.command + "-" + new String(message.payload));
    // 启动后台进程
    handleInThread.start();
    handleOutThread.start();
    startListener.interrupt();
  }

  private Stream open(String destination, boolean canMultipleSend) throws InterruptedException {
    int localId = localIdPool++ * (canMultipleSend ? 1 : -1);
    sendBuffer.write(AdbProtocol.generateOpen(localId, destination));
    Stream stream;
    do {
      synchronized (this) {
        wait();
      }
      stream = connectionStreams.get(localId);
    } while (stream == null);
    return stream;
  }

  public void pushFile(InputStream file, String remotePath) throws Exception {
    // 打开链接
    Stream stream = open("sync:", false);
    // 发送信令，建立push通道
    String sendString = remotePath + ",33206";
    byte[] bytes = sendString.getBytes();
    ByteBuffer send = ByteBuffer.allocate(bytes.length + 8);
    send.put(AdbProtocol.generatePushPacket("SEND", sendString.length()));
    send.put(bytes);
    send.flip();
    stream.write(send);
    // 发送文件
    byte[] byteArray = new byte[4096 - 8];
    int len = file.read(byteArray, 0, byteArray.length);
    do {
      ByteBuffer data = ByteBuffer.allocate(len + 8);
      data.put(AdbProtocol.generatePushPacket("DATA", len));
      data.put(byteArray, 0, len);
      data.flip();
      stream.write(data);
      len = file.read(byteArray, 0, byteArray.length);
    } while (len > 0);
    file.close();
    // 传输完成
    stream.write(AdbProtocol.generatePushPacket("DONE", (int) System.currentTimeMillis()));
    stream.write(AdbProtocol.generatePushPacket("QUIT", 0));
    do {
      synchronized (this) {
        wait();
      }
    } while (!stream.isClosed());
  }

  public String runAdbCmd(String cmd, boolean waitOutput) throws Exception {
    if (waitOutput) {
      Stream stream = open("shell:" + cmd, true);
      do {
        synchronized (this) {
          wait();
        }
      } while (!stream.isClosed());
      return new String(stream.readAllBytes().array());
    } else {
      Stream stream = open("shell:", true);
      stream.write(ByteBuffer.wrap((cmd + "\n").getBytes()));
      return "";
    }
  }

  public Stream tcpForward(int port) throws IOException, InterruptedException {
    Stream stream = open("tcp:" + port, true);
    if (stream.isClosed()) throw new IOException("error forward");
    return stream;
  }

  public Stream localSocketForward(String socketName) throws IOException, InterruptedException {
    Stream stream = open("localabstract:" + socketName, true);
    if (stream.isClosed()) throw new IOException("error forward");
    return stream;
  }

  private void handleIn() {
    try {
      while (!Thread.interrupted()) {
        AdbProtocol.AdbMessage message = AdbProtocol.AdbMessage.parseAdbMessage(channel);
        Stream stream = connectionStreams.get(message.arg1);
        boolean isNeedNotify = stream == null;
        // 新连接
        if (isNeedNotify) {
          stream = createNewStream(message.arg1, message.arg0, message.arg1 > 0);
          connectionStreams.put(message.arg1, stream);
        }
        switch (message.command) {
          case AdbProtocol.CMD_OKAY:
            stream.setCanWrite(true);
            break;
          case AdbProtocol.CMD_WRTE:
            sendBuffer.write(AdbProtocol.generateOkay(message.arg1, message.arg0));
            stream.pushSource(message.payload);
            // sendMoreOk
            sendBuffer.write(AdbProtocol.generateOkay(message.arg1, message.arg0));
            break;
          case AdbProtocol.CMD_CLSE:
            stream.close();
            isNeedNotify = true;
            break;
        }
        if (isNeedNotify) {
          synchronized (this) {
            notifyAll();
          }
        }
      }
    } catch (Exception ignored) {
      close();
    }
  }

  private void handleOut() {
    try {
      while (!Thread.interrupted()) {
        channel.write(sendBuffer.readNext().array());
        if (!sendBuffer.isEmpty()) channel.write(sendBuffer.read(sendBuffer.getSize()).array());
        channel.flush();
      }
    } catch (IOException | InterruptedException ignored) {
      close();
    }
  }

  private Stream createNewStream(int localId, int remoteId, boolean canMultipleSend) {
    return new Stream(false, canMultipleSend, new Stream.UnderlySocketFunction() {
      @Override
      public void write(ByteBuffer buffer) {
        sendBuffer.write(AdbProtocol.generateWrite(localId, remoteId, buffer.array()));
      }

      @Override
      public void close() {
        sendBuffer.write(AdbProtocol.generateClose(localId, remoteId));
      }
    });
  }

  public void close() {
    handleInThread.interrupt();
    handleOutThread.interrupt();
    channel.close();
  }

}

File: scrcpy_android/app/src/main/java/top/saymzx/scrcpy/adb/Adb.kt
/*
 * 本项目为适用于安卓的ADB库，本项目大量借鉴学习了开源ADB库：Dadb，在此对该项目表示感谢
 */

package top.saymzx.scrcpy.adb

import android.util.Log
import okhttp3.internal.notifyAll
import okhttp3.internal.wait
import okio.Buffer
import okio.sink
import okio.source
import java.io.IOException
import java.io.InputStream
import java.net.Socket
import java.nio.charset.StandardCharsets
import java.util.Random

class Adb(host: String, port: Int, keyPair: AdbKeyPair) {

  private val socket: Socket
  private val adbWriter: AdbWriter
  private val adbReader: AdbReader
  private val random = Random()
  private val connectionStreams = HashMap<Int, AdbStream>()
  private var defaultShellStream: AdbStream? = null

  init {
    // 连接socket
    socket = Socket(host, port)
    // 读写工具
    adbReader = AdbReader(socket.source())
    adbWriter = AdbWriter(socket.sink())
    // 连接ADB
    adbWriter.writeConnect()
    var message = adbReader.readMessage()
    // 身份认证
    if (message.command == Constants.CMD_AUTH) {
      adbWriter.writeAuth(Constants.AUTH_TYPE_SIGNATURE, keyPair.signPayload(message))
      message = adbReader.readMessage()
      if (message.command == Constants.CMD_AUTH) {
        adbWriter.writeAuth(Constants.AUTH_TYPE_RSA_PUBLIC, keyPair.publicKeyBytes)
        message = adbReader.readMessage()
      }
    }
    // 认证失败
    if (message.command != Constants.CMD_CNXN) throw IOException("Connection failed: $message")
    // 最大承载长度
    adbWriter.maxPayloadSize = message.arg1
    // 开始读写线程
    val thread = Thread { readAdbThread() }
    thread.priority = Thread.MAX_PRIORITY
    thread.start()
  }

  // 打开一个连接流
  fun open(destination: String, isNeedSource: Boolean): AdbStream {
    val localId = random.nextInt()
    adbWriter.writeOpen(localId, destination)
    val stream = AdbStream(localId, adbWriter, isNeedSource)
    connectionStreams[localId] = stream
    synchronized(stream) {
      stream.wait()
    }
    if (stream.status == -1) throw IOException("连接错误")
    else return stream
  }

  fun pushFile(file: InputStream, remotePath: String) {
    val pushStream = open("sync:start a SYNC service", true)
    val sendString = "$remotePath,33206"
    pushStream.write(createSendPacket("SEND", sendString.length) + sendString.toByteArray())
    val byteArray = ByteArray(60000)
    var len = file.read(byteArray, 0, byteArray.size)
    do {
      var tmpByteArray = createSendPacket("DATA", len) + byteArray.sliceArray(0 until len)
      if (len < byteArray.size) {
        tmpByteArray += createSendPacket("DONE", System.currentTimeMillis().toInt())
      }
      pushStream.write(tmpByteArray)
      len = file.read(byteArray, 0, byteArray.size)
    } while (len > 0)
    pushStream.write(createSendPacket("QUIT", 0))
    synchronized(pushStream) {
      pushStream.wait()
    }
  }

  private fun createSendPacket(id: String, arg: Int): ByteArray {
    val tmpBuffer = Buffer()
    tmpBuffer.clear()
    tmpBuffer.writeString(id, StandardCharsets.UTF_8)
    tmpBuffer.writeIntLe(arg)
    return tmpBuffer.readByteArray()
  }

  fun runAdbCmd(cmd: String, isNeedOutput: Boolean): String {
    if (!isNeedOutput) {
      if (defaultShellStream == null) defaultShellStream = open("shell:", false)
      defaultShellStream!!.write((cmd + "\n").toByteArray())
      return ""
    } else {
      val stream = open("shell:$cmd", true)
      // 等待8秒
      for (i in 0..200) {
        if (stream.status == -1) {
          return stream.source.readUtf8()
        } else {
          Thread.sleep(40)
        }
      }
      return "命令运行超过8秒，未获取命令输出"
    }
  }

  fun tcpForward(port: Int, isNeedSource: Boolean): AdbStream = open("tcp:$port", isNeedSource)

  fun localSocketForward(socketName: String, isNeedSource: Boolean): AdbStream =
    open("localabstract:$socketName", isNeedSource)

  // 读取线程
  private fun readAdbThread() {
    try {
      while (true) {
        val message = adbReader.readMessage()
        val stream = connectionStreams[message.arg1]!!
        when (message.command) {
          Constants.CMD_OKAY -> {
            stream.status = 2
            // 连接成功
            if (stream.remoteId == 0) {
              stream.remoteId = message.arg0
              synchronized(stream) {
                stream.notifyAll()
              }
            }
            synchronized(stream.canWrite) { stream.canWrite.notify() }
          }

          Constants.CMD_WRTE -> {
            adbWriter.writeOkay(message.arg1, message.arg0)
            stream.pushToSource(message.payload)
          }

          Constants.CMD_CLSE -> {
            // 连接断开
            stream.status = -1
            synchronized(stream) {
              stream.notifyAll()
            }
          }
        }
      }
    } catch (_: Exception) {
      close()
    }
  }

  fun close() {
    try {
      defaultShellStream?.close()
      adbReader.close()
      adbWriter.close()
      socket.close()
    } catch (_: Exception) {
    }
  }
}